Improve transparency on macOS

madsmtm opened the pull request:
Extension upon https://github.com/alacritty/alacritty/pull/7928, see that for discussion. I've improved a bit on the window background, by setting the alpha channel to follow the window opacity.

When `window.opacity < 1`, we do two things:
- Set the title bar as transparent (new).
- Disable window border shadows (existing behaviour).

chrisduerr commented on alacritty/src/display/window.rs:
```diff
@@ -182,8 +183,11 @@ impl Window {
 
-        // Set initial transparency hint.
-        window.set_transparent(config.window_opacity() < 1.);
+        // Set initial background opacity and transparency
+        #[cfg(target_os = "macos")]
+        set_titlebar_transparent(&window, config.window_opacity() < 1.);
+        #[cfg(target_os = "macos")]
+        set_background_opacity(&window, config.window_opacity() as _);
```
You're not calling `window.set_transparent` here anymore, why is that?

madsmtm replied:
Because it's called in `WindowBuilder::with_transparent` instead.

chrisduerr replied:
But only on macOS. Contrary to Apple's belief, there are other operating systems. :)

chrisduerr replied:
That was also called before though? Are you suggesting a winit update has removed the necessity for this on all platforms?

madsmtm replied:
Yes, `WindowBuilder::with_transparent` nowadays does the exact same thing as `window.set_transparent` (call `setOpaque:`). In the past, there was confusion about this, because `WindowBuilder::with_transparent` set the window background as well (but it no longer does since https://github.com/rust-windowing/winit/pull/3657).

chrisduerr commented on alacritty/src/display/window.rs:
```diff
@@ -345,6 +349,13 @@ impl Window {
 
     pub fn set_transparent(&self, transparent: bool) {
         self.window.set_transparent(transparent);
+        #[cfg(target_os = "macos")]
+        set_titlebar_transparent(&self.window, transparent);
+    }
+
+    pub fn set_background_opacity(&self, _opacity: f32) {
+        #[cfg(target_os = "macos")]
+        set_background_opacity(&self.window, _opacity as _)
     }
```
Is there ever a scenario where you'd want to call `set_background_opacity` without setting transparency? Seems like the two functions can just be combined.

madsmtm replied:
They probably could be merged, but I think it makes sense to keep them as separate function to self-document specifically what it is that they do.

chrisduerr commented on alacritty/src/display/window.rs:
```diff
@@ -496,3 +507,33 @@ fn use_srgb_color_space(window: &WinitWindow) {
+        _ => return,
+    };
+
+    // If transparent, fill background with NSColor.clearColor,
+    // otherwise fill with NSColor.windowBackgroundColor so that MacOS
+    // properly draws the window's border and opaque title bar.
```
This sounds like the window's background is filled, but surely that's not what is happening, right? Just the titlebar background?

chrisduerr commented on alacritty/src/display/window.rs:
```diff
@@ -496,3 +507,33 @@ fn use_srgb_color_space(window: &WinitWindow) {
 }
+
+#[cfg(target_os = "macos")]
+fn set_background_opacity(window: &WinitWindow, opacity: CGFloat) {
```
These macOS-specific functions should indicate their macOS-specificity. They're also lacking function documentation.

chrisduerr commented on alacritty/src/display/window.rs:
```diff
@@ -167,7 +168,7 @@ impl Window {
             .with_theme(config.window.theme())
             .with_visible(false)
-            .with_transparent(true)
+            .with_transparent(config.window_opacity() < 1.)
```
I believe we've had it this way in the past and that changed. Are you sure this is safe on all platforms? Because I'm not sure that this doesn't change Alacritty's behavior.

madsmtm replied:
You're right that the behaviour here is different on X11 and Windows, didn't consider other platforms when making this. Have reverted the change.

chrisduerr commented on alacritty/src/display/window.rs:
```diff
@@ -182,8 +183,11 @@ impl Window {
 
-        // Set initial transparency hint.
-        window.set_transparent(config.window_opacity() < 1.);
+        // Set initial background opacity and transparency
```
```suggestion
        // Set initial background opacity and transparency.
```

chrisduerr reviewed (changes requested):
Missing changelog entry.

sfsam commented:
I'm seeing the same thing I observed in the discussion of #7928: No borders and a white title bar even though `decorations_theme_variant = "Dark"` is set. I think this is due to something that was changed in the move to winit 0.3.0. I'm on MacOS Sonoma 14.4.1. FWIW, my window config:
```
[window]
decorations_theme_variant = "Dark"
decorations = "full"
option_as_alt = "OnlyLeft"
resize_increments = true
dimensions = { columns = 100, lines = 24 }
```

nixpulvis commented:
> Set the title bar as transparent (new)

Seems to me this may be undesired. I hate to think we need yet another macOS specific configuration for it though.

kchibisov commented:
The end window opacity is not as it should be, it's just generally more _opaque_ compared to what it was before or glitches.

madsmtm commented:
I've updated the PR and reverted some of the changes I probably mistakenly thought were for the better, though I'm too unsure of what the desired behaviour is (I don't like transparent windows myself, and don't use Alacritty).

The first two commits should bring Alacritty back in line with what it was pre winit 0.30.

Would be happy if someone else would take over for the rest.

nixpulvis commented:
I should be able to finish testing and implementing this PR, though my macOS API knowledge is limited and it may take me a little while to get around to everything. Right now it seems the main things that need changing are with respect to the titlebar and fixing whatever is causing #7969. At first I didn't notice that issue myself, but now transparency is broken for me as well, without any changes to the code or configuration, so it seems some macOS intermittent issue. Fun.

I assume the best way to do this would be to make a new PR with the current state of this one and push to that myself as I make progress?

chrisduerr commented:
> I assume the best way to do this would be to make a new PR with the current state of this one and push to that myself as I make progress?

Probably easiest this way, yes.

nixpulvis commented:
So after catching up on this issue, it seems like https://github.com/rust-windowing/winit/pull/3722 fixes the main issue (#7969), so what's left is to figure out what we actually want, if anything.

I'm unsure that having a transparent titlebar is the right default because it can be achieved with `window.decorations = "Transparent"` (however, I still think that could use with some improvement so the content area doesn't overlap with the titlebar buttons, but that's an existing concern).

I also don't know if it's better to have drop shadows or not with a transparent background. My instinct would be to have them, but given that we don't have them already, I'm fine leaving it as is.

Looking back at #7928, it seems like everything there has been addressed.

Here's a screenshot of an alacritty window (`opacity = 0.75`) from master with winit pointing to https://github.com/rust-windowing/winit/pull/3722:

<img width="280" alt="Screenshot 2024-06-10 at 10 35 02 AM" src="https://github.com/alacritty/alacritty/assets/715947/3bd76bfb-1f07-40b5-9b1d-248079643906">

and here's one without `opacity` set:

<img width="337" alt="Screenshot 2024-06-10 at 10 40 00 AM" src="https://github.com/alacritty/alacritty/assets/715947/478d0f4a-1eee-4c70-bbd6-05d4378adcb5">

Yes the transparent window's titlebar is a less dark color, but I don't see the red bleeding into it like it appears to be in the original PR and I believe this behavior looks reasonable.

madsmtm commented:
I'll note another option: Use `-[NSWindow setAlphaValue:]`, this also sets the opacity of all the decorations (the border and the title bar) (in this mode, the application should internally render as-if it was opaque)

nixpulvis commented:
@madsmtm I was looking at that the other day. Might be a better solution in that it might look better, but I'm not sure it's worth platform specific code if `winit` still implements transparency the other way. We don't want to change how things work on Linux/Windows.

kchibisov commented:
@sfsam I'd suggest to bring it upstream if after the https://github.com/alacritty/alacritty/pull/8043 the issue will be still the same, because I can not repro it myself and everything works as expected(or before).

It's true that internally winit touched a lot of code on macOS, so there's a chance that with 14+ something is not working, given that I can only test version 12.

kchibisov commented:
Closing in favor of https://github.com/alacritty/alacritty/pull/8043
